File: binarytreeverticalordertraversal.js
//Objective is to do a vertical order traversal of a binary tree.


//O(n) solution that uses BFS and a queue. Every time we go left, the 'level'
//decrements, and every time we go right, the 'level' increments.

if (!root) {
    return []
}

let map = new Map()
let queue = []
queue.push({node: root, level: 0})

//Keep track of the lowest and highest level for iteration later
let min = 0
let max = 0

while (queue.length > 0) {
    let curr = queue.shift() 
    
    if (!map.has(curr.level)) {
        map.set(curr.level, [])
    }
    map.get(curr.level).push(curr.node.val)
    
    min = Math.min(min, curr.level)
    max = Math.max(max, curr.level)
    
    if (curr.node.left) {
        queue.push({node: curr.node.left, level: curr.level - 1})
    }
    
    if (curr.node.right) {
        queue.push({node: curr.node.right, level: curr.level + 1})
    }
}

let result = []

//We do 'i - min' here since 'i' might be negative
for (let i = min; i <= max; i++) {
    result[i - min] = map.get(i)
}

return result

File: binarytreeverticalordertraversal.java
class Solution {
    public List<List<Integer>> verticalOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        
        Map<Integer, ArrayList<Integer>> map = new HashMap<>();
        LinkedList<Pair<TreeNode, Integer>> stack = new LinkedList<>();
        stack.add(new Pair(root, 0));
        int min = Integer.MAX_VALUE;
        int max = Integer.MIN_VALUE;
        
        while (!stack.isEmpty()) {
            Pair<TreeNode, Integer> curr = stack.poll();
            TreeNode node = curr.getKey();
            int level = curr.getValue();
            
            if (!map.containsKey(level)) {
                map.put(level, new ArrayList<>());
            }
            map.get(level).add(node.val);
            
            min = Math.min(min, level);
            max = Math.max(max, level);
            
            if (node.left != null) {
                stack.add(new Pair(node.left, level - 1));
            }
            
            if (node.right != null) {
                stack.add(new Pair(node.right, level + 1));
            }
        }
        
        for (int i = min; i <= max; i++) {
            result.add(map.get(i));
        }
        
        return result;
    }
}